platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/isolation/IsolatedActionSerializer.kt

```diff
@@ -0,0 +1,186 @@
+     * These might include references to classes, value sources and build services.
+     *
+     * Maps the integer written to the serialized [graph] to the external reference.
+     * See [EnvironmentEncoder] and [EnvironmentDecoder] for details.
+     */
+    val environment: Map<Int, Any>,
+)
+
+
+internal
+class IsolatedActionSerializer(
+    private val owner: IsolateOwner,
+    private val beanStateWriterLookup: BeanStateWriterLookup,
+    private val isolatedActionCodecs: IsolatedActionCodecsFactory
+) {
+    fun <G : Any> serialize(action: G): SerializedIsolatedActionGraph<G> {
+        val outputStream = ByteArrayOutputStream()
+        val environmentEncoder = EnvironmentEncoder()
+        serializeTo(outputStream, environmentEncoder, action)
+        return SerializedIsolatedActionGraph(
+            outputStream.toByteArray(),
+            environmentEncoder.toLookup()
+        )
+    }
+
+    private
+    fun serializeTo(
+        outputStream: ByteArrayOutputStream,
+        environmentEncoder: EnvironmentEncoder,
+        action: Any
+    ) {
+        writeContextFor(outputStream, environmentEncoder).useToRun {
+            runWriteOperation {
+                withIsolate(owner) {
+                    write(action)
+                }
+            }
+        }
+    }
+
+    private
+    fun writeContextFor(
+        outputStream: OutputStream,
+        classEncoder: ClassEncoder,
+    ) = DefaultWriteContext(
+        codec = isolatedActionCodecs.isolatedActionCodecs(),
+        encoder = KryoBackedEncoder(outputStream),
+        beanStateWriterLookup = beanStateWriterLookup,
+        logger = logger,
+        tracer = null,
+        problemsListener = ThrowingProblemsListener,
+        classEncoder = classEncoder
+    )
+}
+
+
+internal
+class IsolatedActionDeserializer(
+    private val owner: IsolateOwner,
+    private val beanStateReaderLookup: BeanStateReaderLookup,
+    private val isolatedActionCodecs: IsolatedActionCodecsFactory
+) {
+    fun <G : Any> deserialize(action: SerializedIsolatedActionGraph<G>): G =
+        readerContextFor(action).useToRun {
+            runReadOperation {
+                withIsolate(owner) {
+                    readNonNull()
+                }
+            }
+        }
+
+    private
+    fun readerContextFor(
+        action: SerializedIsolatedActionGraph<*>
+    ) = DefaultReadContext(
+        codec = isolatedActionCodecs.isolatedActionCodecs(),
+        decoder = KryoBackedDecoder(action.graph.inputStream()),
+        beanStateReaderLookup = beanStateReaderLookup,
+        logger = logger,
+        problemsListener = ThrowingProblemsListener,
+        classDecoder = EnvironmentDecoder(action.environment)
+    )
+}
+
+
+private
+class EnvironmentEncoder : ClassEncoder {
+
+    private
+    val refs = IdentityHashMap<Class<*>, Int>()
+
+    override fun WriteContext.encodeClass(type: Class<*>) {
+        val existing = refs[type]
```

Comment:
💅 

```
val id = refs.computeIfAbsent(type) { _ -> refs.size }
writeSmallInt(id)
```

Reply:
That code was moved and I'd rather not change it further (given the potential performance implications for such a low level operation).

Reply:
Actually, it makes sense to change it since this is a new code path (at first I thought this was still used by CC serialization). Thanks!